### PR TITLE
bugfix discoverer & add discovery cache & bugfix agent index count & add some feature

### PR DESCRIPTION
**Pull request by kpango**
**Signed-off-by: kpango <i.can.feel.gravity@gmail.com>**

##### bugfix
- discoverer returns the same pointer of grpc payload it makes problem so I fixed about that
- agent reports index count using non-atomic counter it makes the wrong value, so I fixed that using atomic variable
- agent crashes when observability is nil

##### feature
- discoverer now caches each response
- internal Cassandra now supports tls authentication and completely support DC aware routing
- grpc client now supports rebalanced reconnecting
- grpc client now supports DNS optimized balanced connection pooling
- internal/net package now overrides official net package
- index manager now supports custom creation pool size

##### prototype
- replication manager controller
- replication manager agent
- internal compressor client

### Environment:

- Golang Version: 1.14
- Docker Version: 19.03.6-rc2
- Kubernetes Version: 1.17.4
- NGT Version: 1.9.1

 

- [ ] Bug fix [type/bug]
- [x] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -212,7 +212,8 @@ deps: \
 
 .PHONY: goimports/install
 goimports/install:
-	GO111MODULE=off go get -u golang.org/x/tools/cmd/goimports
+	go get -u golang.org/x/tools/cmd/goimports
```
Comment (hlts2) [inline Makefile:215]: if we delete `GO111MODULE=off `, `go.mod` of root directory may change. 🤔

File: cmd/manager/replication/agent/main.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+				}
+				return cfg, &cfg.GlobalConfig, nil
+			}),
+			runner.WithDaemonInitializer(func(cfg interface{}) (runner.Runner, error) {
+				return usecase.New(cfg.(*config.Data))
+			}),
+		)
+	})(); err != nil {
+		log.Fatal(err, info.Get())
+		return
```
Comment (hlts2) [inline cmd/manager/replication/agent/main.go:56]: `log.Fatal` exits program, so I think that `return` is not needed 🤔

File: internal/net/grpc/pool.go [hunk 3]
```diff
@@ -45,24 +67,104 @@ type ClientConnPool struct {
...
+	cp, err := newPool(ctx, host, port, size, dopts...)
+	if err != nil {
+		if cp != nil {
+			err = errors.Wrap(cp.disconnect(), err.Error())
+		}
+		port, ok = scanGRPCPort(ctx, host)
+		if !ok {
+			return nil, err
+		}
+		cp, err = newPool(ctx, host, port, size, dopts...)
```
Comment (rinx) [inline internal/net/grpc/pool.go:92]: badge
the `err` should be handled.

File: Makefile (hunk 1 above)
Comment (kpango) [reply Makefile:215]: maybe it's okay, if developer already have goimports this section never run

File: cmd/manager/replication/agent/main.go (hunk 2 above)
Comment (kpango) [reply cmd/manager/replication/agent/main.go:56]: it's the same as others, log.Fatal is not always include os.Exit it's depends on library, we shouldn't trust them

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (kpango) [reply Makefile:215]: yes, recently goimports becomes compatible with go module, so I delete GO111MODULE

File: internal/config/grpc.go [hunk 4]
```diff
@@ -25,13 +25,13 @@ import (
...
-	Addrs               []string    `json:"addrs" yaml:"addrs"`
-	HealthCheckDuration string      `json:"health_check_duration" yaml:"health_check_duration"`
-	ConnectionPool      int         `json:"connection_pool" yaml:"connection_pool"`
-	Backoff             *Backoff    `json:"backoff" yaml:"backoff"`
-	CallOption          *CallOption `json:"call_option" yaml:"call_option"`
-	DialOption          *DialOption `json:"dial_option" yaml:"dial_option"`
-	TLS                 *TLS        `json:"tls" yaml:"tls"`
+	Addrs               []string        `json:"addrs" yaml:"addrs"`
+	HealthCheckDuration string          `json:"health_check_duration" yaml:"health_check_duration"`
+	ConnectionPool      *ConnectionPool `json:"connection_pool" yaml:"connection_pool"`
```
[the remark this answers is not in the document]
Comment (kpango) [reply internal/config/grpc.go:30]: oh, I forgot  about that I'll fix them soon

File: internal/net/grpc/pool.go (hunk 3 above)
Comment (kpango) [reply internal/net/grpc/pool.go:92]: I forgot about that I'll fix them soon

Comment (kpango) [reply internal/net/grpc/pool.go:92]: fixed

File: internal/config/grpc.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kpango) [reply internal/config/grpc.go:30]: fixed

File: charts/vald/values.yaml [hunk 5]
```diff
@@ -271,6 +271,13 @@ defaults:
       connection_pool: 3
       # defaults.grpc.client.health_check_duration -- gRPC client health check duration
       health_check_duration: "1s"
+      connection_pool:
+        # defaults.grpc.client.connection_pool.enable_rebalance -- enables gRPC client connection pool rebalance
+        enable_rebalance: true
+        # defaults.grpc.client.connection_pool.enable_rebalance -- gRPC client connection pool rebalance duration
```
Comment (rinx) [inline charts/vald/values.yaml:277]: badge
```suggestion
        # defaults.grpc.client.connection_pool.rebalance_duration-- gRPC client connection pool rebalance duration
```

File: charts/vald/values.yaml [hunk 6]
```diff
@@ -271,6 +271,13 @@ defaults:
...
+        rebalance_duration: 30m
+        # defaults.grpc.client.connection_pool.enable_rebalance -- gRPC client connection pool size
```
Comment (rinx) [inline charts/vald/values.yaml:279]: badge
```suggestion
        # defaults.grpc.client.connection_pool.size-- gRPC client connection pool size
```

File: internal/net/net.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+				mu.Lock()
+				ports = append(ports, port)
+				mu.Unlock()
+				return conn.Close()
+			}
+		})
+	}
+	err = eg.Wait()
+
+	return ports, nil
```
Comment (hlts2) [inline internal/net/net.go:115]: Even if `eg.Wait` returns error,  errors are thrown away 🤔

**Pull request merged**